plat/common/paging.c

```diff
@@ -0,0 +1,1320 @@
+{
+	__paddr_t pt_paddr;
+
+	UK_ASSERT(level < PT_LEVELS);
+
+	pt_paddr = pgarch_pt_unmap(pt, pt_vaddr, level);
+	UK_ASSERT(pt_paddr != __PADDR_INV);
+
+	pg_ffree(pt, pt_paddr, PAGE_LEVEL);
+
+#ifdef CONFIG_PAGING_STATS
+	UK_ASSERT(pt->nr_pt_pages[level] > 0);
+	pt->nr_pt_pages[level]--;
+#endif /* CONFIG_PAGING_STATS */
+}
+
+static inline int pg_largest_level(__vaddr_t vaddr, __paddr_t paddr, __sz len,
+				   unsigned int max_lvl)
+{
+	unsigned int lvl = max_lvl;
+
+	while (lvl > PAGE_LEVEL) {
+		if (PAGE_Lx_HAS(lvl) &&
+		    PAGE_Lx_ALIGNED(vaddr, lvl) &&
+		    PAGE_Lx_ALIGNED(paddr, lvl) &&
+		    PAGE_Lx_SIZE(lvl) <= len)
+			return lvl;
+
+		lvl--;
+	}
+
+	return lvl;
+}
+
+static int pg_page_map(struct uk_pagetable *pt, __vaddr_t pt_vaddr,
+		       unsigned int level, __vaddr_t vaddr, __paddr_t paddr,
+		       __sz len, unsigned long prot, unsigned long flags,
+		       __pte_t template, unsigned int template_level)
+{
+	unsigned int to_lvl = PAGE_FLAG_SIZE_TO_LEVEL(flags);
+	unsigned int max_lvl = pg_page_largest_level;
+	unsigned int lvl = level;
+	unsigned int tmp_lvl;
+	__vaddr_t pt_vaddr_cache[PT_LEVELS];
+	__paddr_t pt_paddr;
+	__pte_t pte;
+	__sz page_size;
+	unsigned int pte_idx;
+	int rc, alloc_pmem;
+
+	UK_ASSERT(len > 0);
+	UK_ASSERT(PAGE_Lx_ALIGNED(len, to_lvl));
+	UK_ASSERT(PAGE_Lx_ALIGNED(vaddr, to_lvl));
+	UK_ASSERT(vaddr <= __VADDR_MAX - len);
+	UK_ASSERT(ukarch_vaddr_range_isvalid(vaddr, vaddr + len));
+
+	if (paddr != __PADDR_ANY) {
+		UK_ASSERT(PAGE_Lx_ALIGNED(paddr, to_lvl));
+		UK_ASSERT(paddr <= __PADDR_MAX - len);
+		UK_ASSERT(ukarch_paddr_range_isvalid(paddr, paddr + len));
+
+		alloc_pmem = 0;
+	} else
+		alloc_pmem = 1;
+
+	if (!(flags & PAGE_FLAG_FORCE_SIZE)) {
+		if (level < max_lvl)
+			max_lvl = level;
+
+		to_lvl = pg_largest_level(vaddr, paddr, len, max_lvl);
+	}
+
+	UK_ASSERT(lvl >= to_lvl);
+	pt_vaddr_cache[lvl] = pt_vaddr;
+
+	pte_idx = PT_Lx_IDX(vaddr, lvl);
+	page_size = PAGE_Lx_SIZE(lvl);
+	do {
+		/* This loop is responsible for walking the page table down
+		 * until we reach the desired level. If there is a page table
+		 * missing on the way it is allocated and linked.
+		 */
+		while (lvl > to_lvl) {
+			/* We are too high and need to walk further down */
+			rc = ukarch_pte_read(pt_vaddr, lvl, pte_idx, &pte);
+			if (unlikely(rc))
+				return rc;
+
+			if (PT_Lx_PTE_PRESENT(pte, lvl)) {
+				/* If there is already a larger page mapped
+				 * at this address return an error.
+				 */
+				if (PAGE_Lx_IS(pte, lvl))
+					return -EEXIST;
+
+				pt_vaddr = pgarch_pt_pte_to_vaddr(pt, pte, lvl);
+			} else {
+				/* There is nothing here, not even a page table.
+				 * So allocate a new one and link it.
+				 */
+				rc = pg_pt_alloc(pt, &pt_vaddr, &pt_paddr,
+						 lvl - 1);
+				if (unlikely(rc))
+					return rc;
+
+				if (!(flags & PAGE_FLAG_KEEP_PTES))
+					pte = template;
+				else
+					template_level = lvl;
+
+				pte = pgarch_pt_pte_create(pt, pt_paddr, lvl,
```

Comment:
Maybe I'm missing something here, but how does `pgarch_pte_create()` know that this PTE points to a table rather than an actual mapping? The x86 implementation sets the PSE bit conditionally to `level  > PAGE_LEVEL`, which implies that `pgarch_pte_create` only works PTEs that describe mappings.

Reply:
Not sure what you mean. I differentiate in the code between `pgarch_pt_pte_create` which is for page tables and `pgarch_pte_create` which is for data pages. The code that you commented creates a page table PTE so it uses the `pt_pte` variant.

Reply:
Argh, I didn't notice that there are actually two functions, the names are too similar 😖 You're right, there are no issues there, sorry about the confusion.